Updated the FastLinAlg package

kschwede opened the pull request:
In particular, this adds the isCodimAtLeast method, improves documetnation and fixes some minor bugs.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -15,7 +15,7 @@ Version => "0.1", Date => "October 21st, 2019", Authors => {
     }
 }, --this file is in the public domain
-Headline => "faster linear algebra operations")
+Headline => "A package for faster linear algebra operations.", DebuggingMode => true, Reload=>true)
```
Please leave DebuggingMode and Reload off.

kschwede replied:
I've done this one, I'll reply as I've done the others.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -77,7 +98,7 @@ StrategyDefault := new HashTable from {
 --protect nanosleep;
 --if (not isGlobalSymbol "nanosleep") then protect getSymbol("nanosleep");
 
-StrategyCurrent = new MutableHashTable from {LexLargest=>0, LexSmallestTerm => 50, LexSmallest=>50, GRevLexSmallestTerm=>0, GRevLexSmallest=>0, GRevLexLargest=>0,Random=>0,RandomNonzero=>0};
```
Here you intend to override some values from `StrategyDefault`, so it would be good to write 
```
StrategyCurrent = StrategyDefault ++ {LexLargest=>0, LexSmallestTerm => 50, LexSmallest=>50, GRevLexSmallestTerm=>0, GRevLexSmallest=>0, GRevLexLargest=>0,Random=>0,RandomNonzero=>0};
```
, for you could omit from the list those items with the same values.

kschwede replied:
I think you are looking at the replaced code.  The current code creates a Mutable strategy, based on the default strategy, for the user to play with.  There may be other ways to accomplish this, but this seemed reasonable.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -612,6 +649,19 @@ RnReductionP(ZZ, Ring, ZZ):= opts -> (n1, R1, p)-> (
 
+verifyStrategy := (passedStrat) -> (
+    if not ((class passedStrat === HashTable) or ((class passedStrat) === MutableHashTable)) then return false;
+    if not (passedStrat #? LexLargest) then return false;
```
Rewrite using the logical operator `and`, without `if` or `return`.

kschwede replied:
I believe I've done this.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -710,10 +760,14 @@ Rn(ZZ, Ring) := opts -> (n1, R1) -> (
+        numberOfMinorsCompute = opts.MaxMinors;)
     else (
-        numberOfMinorsCompute = opts.MaxMinors;);
+        numberOfMinorsCompute = 5*minNumberToCutDown + 8*ceiling(log_1.3(possibleMinors));
```
At this point you should signal an error if it isn't `null`.  (Users can make mistakes.)

kschwede replied:
I think you are misreading this.  The user can pass a function, a number, or anything else.  If they pass anything else, it reverts to a default function.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -781,10 +844,11 @@ chooseGoodMinors(ZZ, ZZ, Matrix) := opts -> (howMany, minorSize, M1) -> (
     R2 := reorderPolynomialRing(GRevLex, ambR);
     f := map(R2, ambR);
     sumMinors := trim ideal(sub(0, ambR));
+    --1/0;
```
This doesn't look useful.

kschwede replied:
I've removed the commented out --1/0s, throughout.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -799,6 +863,10 @@ chooseGoodMinors(ZZ, ZZ, Matrix) := opts -> (howMany, minorSize, M1) -> (
             );
             i = i+1;
         );
+        if not (opts.PeriodicCheckFunction === null) then (
```
You could use `=!=` here.

kschwede replied:
I believe I've done this throughout.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -812,26 +880,39 @@ chooseGoodMinors(ZZ, ZZ, Matrix) := opts -> (howMany, minorSize, M1) -> (
+        minorsCount = 5*myDim + 2*log_1.3(possibleMinors); );
+--        print class opts.MaxMinors;
+
+--    1/0;
```
This, too.

kschwede replied:
Done.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -855,6 +940,71 @@ projDim(Module) := opts -> (N1) -> (
+isGBDone := (myGB) -> (
+    --a temporary function for finding out if a gb computation is done.
+    myStr := status myGB;
+    if (0 < #select("status: done", myStr)) then return true else return false;
```
@mikestillman -- is there a better way to do this?  If not, we should invent one.

kschwede replied:
I agree.  I don't know a better way to do it though.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -855,6 +940,71 @@ projDim(Module) := opts -> (N1) -> (
+isCodimAtLeast(ZZ, Ideal) := opts -> (n1, I1) -> (
+    R1 := ring I1;
+    S1 := ambient R1;
+    if (not isPolynomialRing(S1)) then error "isCodimAtLeast:  This requires an ideal in a polynomial ring, or in a quotient of a polynomial ring.";
```
Lots of redundant parentheses in these lines (and elsewhere).  Example:
```
if (codim monomialIdeal(I1) >= n1) then return true;
```
could be
```
if codim monomialIdeal I1 >= n1 then return true;
```

kschwede replied:
I actually strongly prefer the parenethesed version.  I find it much easier to read.

DanGrayson replied:
Okay, that's fine.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -855,6 +940,71 @@ projDim(Module) := opts -> (N1) -> (
+        if (isGBDone(myGB)) then i = opts.PairLimit;
+        i = i + 1;
+    );
+    return null;
```
This line can be omitted.

kschwede replied:
I realize that, but my authors and I thought this way would be easeir to read.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -855,6 +940,71 @@ projDim(Module) := opts -> (N1) -> (
+ --   baseDeg := apply(sum(apply(first entries vars S1, t1 -> degree t1)), v -> ceiling(v/vCount)); --use this as the base degree to step by (probably we should use a different value)
+    i := 1;
+    --curLimit := baseDeg;
+    local curLimit;
```
An alternative to declaration with `local` is to change the first assignments below to `:=`.

kschwede replied:
The first assignments appear inside a loop (or at least did initially).  Won't it cause issues that way?  (or am I misremembering).  We don't want to do := assignments repeatedly at each step inside the loop, right?

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -1269,8 +1425,8 @@ doc ///
         Text
-            This function looks at subtrices of the given matrix, and tries to find
-            one of the specified rank.  If it succeeds, it returns a list of two lists.  
+            This function looks at subtmatrices of the given matrix, and tries to find
```
Still spelled wrong!

kschwede replied:
Fixed.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -1299,6 +1455,14 @@ doc ///
         (Rn, ZZ, Ring)
         [Rn, Verbose]
         [Rn, ModP]
```
The name "Rn" is too brief!  Think of something descriptive.

kschwede replied:
RegularInCodimensionN?
That seemed rather long to type.  
I thought we wanted to avoid things like 
isRn
as a matter of style.  Although that would be another option.

I think Rn is the correct and extensively used mathematical name for this though.

mikestillman replied:
How about SerreRn, SerreSn?

kschwede replied:
That's good!  Although I suppose isSerreSn is redundant.  (Since the S is Sn stands for Serre).  I don't think I have any particularly good ideas about how to implement Sn checking though.  There will be another improvement to this package when we get the randomRationalPoints working as well as we want, but that's probably for 1.17.

DanGrayson replied:
"isSerreRn" is bad for two reasons: it contains "R", which is a symbol with no English meaning; it contains "n", which is just an a arbitrary identifier that could have been any other letter of the alphabet.

kschwede replied:
Hi Dan, 
I guess we could just go with 
regularInCodimension
We presumably want to avoid abbreviations like
regInCodim
regularInCodim
On the other hand "Rn" is really the way this is used in mathematical speech I believe.  Specifically, people will really say things like this scheme/variety/ring is R0 / R1 / R2...
Let me know.

mikestillman replied:
I think `regularInCodimension` is a good (self-dcumenting) name.  The documentation could mention the name`Rn`.  I agree that even though it is standard math usage in commutative algebra, it is too short of a name, making it hard to see, search for, etc.  I guess also: the name of the function should then be `R`, not `Rn`, but that isn't allowed either.

kschwede replied:
Ok, I've renamed it.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -1569,8 +1742,8 @@ doc ///
-        R1: ZZ
-           a polynomial ring
+        b: Boolean
+           true if 
```
The thought here was left incomplete.

kschwede replied:
Worse, the function doesn't return a Boolean, this is now fixed.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -1614,6 +1787,80 @@ doc ///
+        [isCodimAtLeast, PairLimit]
+        SPairsFunction
+    Headline
+        returns true if we can quickly see if the codim is at least a given number
```
change second "if" to "whether"

kschwede replied:
Done.

DanGrayson commented:
Thanks!

About StrategyCurrent -- Yes, I didn't understand that in your application the mutability is important.

Re "The user can pass a function, a number, or anything else. If they pass anything else, it reverts to a default function."  The problem is that if they make a mistake, and you regard their value as "something else", then they may not get the desired effect.  It would be better to insist on something specific if they want the default.

About ":=" inside a loop.  No, the declaration of the variable as local inside a loop is a static attribute that is determined when the function is parsed and compiled.  The assignment itself is all that happens at runtime.

Another thing I just noticed : if you check user input (as in `class passedStrat === HashTable`) by insisting that the class of something they give you be equal to a certain thing, then that will prevent them from giving you a class that inherits from that class.  I.e., if you ask for an object of class HashTable, and X is a new type of hashtable, then they can't give you an instance of X.  That may not be important for you, but in almost all cases, the use of "instance" instead will look just as good and allow more flexible use of inheritance.

DanGrayson commented:
How did you make all those replies to my comments appear as subitems of your latest comment?

kschwede commented:
I made comments by clicking a box that looks like a "text-message" icon in the upper right hand corner of my github page, when viewing the pull request.
![WhereIclicked](https://user-images.githubusercontent.com/6293041/82767987-011f7100-9de9-11ea-90bd-0c8bc2b38def.png)
That let me respond individually.

Good point about the MaxMinors => null.  It now gives an error if the user doesn't provide the correct input.  

I was confused about the := in a loop.  I've changed it.

For the other instance vs class thing, good point.  I've changed it.

DanGrayson commented:
Are you done updating it?

kschwede commented:
Yes, I'm done.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -705,15 +754,19 @@ Rn(ZZ, Ring) := opts -> (n1, R1) -> (
     else (
-        numberOfMinorsCompute = opts.MaxMinors;);
+        --numberOfMinorsCompute = 5*minNumberToCutDown + 8*ceiling(log_1.3(possibleMinors));
+        error "regularInCodimension:  You must pass MaxMinors either a function or a number.";
```
Now you don't have a default number (to be used if "null" is there).  Is that intentional?  You could have it.

kschwede replied:
Hi Dan, it is indeed intentional.  The option has a default value (actually, the same as what null used to force).

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -855,6 +936,70 @@ projDim(Module) := opts -> (N1) -> (
+isGBDone := (myGB) -> (
+    --a temporary function for finding out if a gb computation is done.
+    myStr := status myGB;
+    if (0 < #select("status: done", myStr)) then return true else return false;
```
It's a tautology that
```
if X then true else false
```
is equal to X, so the circumlocution is confusing.

kschwede replied:
Fixed.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -855,6 +936,70 @@ projDim(Module) := opts -> (N1) -> (
+isGBDone := (myGB) -> (
+    --a temporary function for finding out if a gb computation is done.
+    myStr := status myGB;
+    if (0 < #select("status: done", myStr)) then return true else return false;
```
It's better to replace
```
0 < #select("status: done", myStr)
```
by
```
match("status: done", myStr)
```

kschwede replied:
Fixed.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -1614,50 +1784,124 @@ doc ///
+    Outputs
+        : Boolean
+           {\tt true} if the codimension of I is at least n
+        : 
```
This second output item conveys the idea that the function returns 2 outputs, rather than the idea that the function returns 1 output, which could be of two types.

kschwede replied:
I see, I guess I don't know what the right way to document that is then.

DanGrayson replied:
Try this:
```
    Outputs
        : Boolean
           {\tt true} or {\tt false}, indicating whether the codimension of I is at least n, if it can be determined, else {\tt null}
```

We cheat a little, in that "null" is not of type Boolean.

kschwede replied:
Ok, I believe I've made this change throughout.

DanGrayson commented on M2/Macaulay2/packages/FastLinAlg.m2:
```diff
@@ -1614,50 +1784,124 @@ doc ///
+        : Boolean
+           {\tt true} if the dimension of I is at most n
+        : 
+            {\tt null} if the function cannot tell whether the dimension is at most n
```
same here

kschwede replied:
Done.
